src/shortcuts/__tests__/undo.ts

```diff
@@ -0,0 +1,295 @@
+    {
+      type: 'existingThoughtChange',
+      context: ['a'],
+      oldValue: 'b',
+      newValue: 'bd',
+      rankInContext: 0,
+      thoughtsRanked: [{ value: 'a', rank: 0 }, { value: 'b', rank: 0 }]
+    },
+    // dispensible set cursor (which only updates datanonce)
+    { type: 'setCursor', thoughtsRanked: null },
+    // undo setCursor and thoughtChange in a sinle action
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+    - b
+    - c
+    - d`
+
+  expect(exported).toEqual(expectedOutput)
+})
+
+it('state remains unchanged if there are no inverse patches', () => {
+  const store = createTestStore()
+
+  store.dispatch(importText(RANKED_ROOT, `
+    - a
+     - b
+     - c
+     - d`)
+  )
+
+  const prevState = store.getState()
+  expect(prevState.inversePatches.length).toEqual(0)
+
+  store.dispatch({ type: 'undoAction' })
+
+  expect(store.getState()).toEqual(prevState)
+})
+
+it('newThought action should be merged with the succeeding patch', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b`
+    ),
+    { type: 'newThought', value: 'c' },
+    { type: 'newThought', value: 'd' },
+    {
+      type: 'existingThoughtChange',
+      context: [ROOT_TOKEN],
+      oldValue: 'd',
+      newValue: 'd1',
+      rankInContext: 3,
+      thoughtsRanked: [
+        {
+          value: 'd',
+          rank: 3,
+        }
+      ]
+    },
+    // undo thought change and preceding newThought action
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+  - b
+  - c`
+
+  expect(exported).toEqual(expectedOutput)
+
+})
+
+it('undo contiguous changes', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - A
+      - B`
+    ),
+    {
+      type: 'existingThoughtChange',
+      newValue: 'Atlantic',
+      oldValue: 'A',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'A', rank: 0 }]
+    },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'Atlantic City',
+      oldValue: 'Atlantic',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'Atlantic', rank: 0 }]
+    },
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - A
+  - B`
+
+  expect(exported).toEqual(expectedOutput)
+
+})
+
+it('state.alert is omitted from the undo patch', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - A
+      - B`
+    ),
+    { type: 'alert', value: 'test' },
+    { type: 'archiveThought' },
+  ])
+
+  expect(store.getState().inversePatches.length).toEqual(0)
+
+})
```

Comment:
That's a valid point. We should be looking into the `inversePatch` to verify `alert` is actually omitted. 
On a side note, the `inversePatches` is empty because the cursor value is null, and hence, the `archiveThought` doesn't update the state. Since there's no change in the state, no patch is generated.
I've updated the test nevertheless, to work as intended.